src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -142,7 +143,42 @@ pub struct StorageControllerState<T, S = mz_stash::Sqlite> {
     /// This collection only grows, although individual collections may be rendered unusable.
     /// This is to prevent the re-binding of identifiers to other descriptions.
     pub(super) collections: BTreeMap<GlobalId, CollectionState<T>>,
-    pub(super) stash: S,
+    pub(super) stash: StorageStash<S>,
+}
+
+/// Storage specific wrapper over a stash that provides type safety for the various types of
+/// collections stored in the inner Stash.
+#[derive(Debug)]
+pub struct StorageStash<S> {
```

Comment:
I think `StashCollectionFactory` might be more clear, but I understand if people hate the word `Factory`